packages/codegen/src/generateMetadataFile.ts

```diff
@@ -59,59 +73,64 @@ function generateFields(config: Config, dbMetadata: EntityDbMetadata): Record<st
       fieldName: "id",
       fieldIdName: undefined,
       required: true,
-      serde: new ${KeySerde}("${dbMetadata.tagName}", "id", "id", "${dbMetadata.primaryKey.columnType}"),
+      serde: new ${KeySerde}({
+        fieldName: "id",
+        columnName: "id",
+        dbType: "${dbMetadata.primaryKey.columnType}",
+        tagName: "${dbMetadata.tagName}"
+      }),
       immutable: true,
     }
   `;
 
-  dbMetadata.primitives.forEach((p) => {
-    const { fieldName, derived, columnName, columnType, superstruct } = p;
+  dbMetadata.primitives.forEach((field) => {
+    const { fieldName, derived, columnName, columnType, superstruct } = field;
     const serdeType = superstruct
-      ? code`new ${SuperstructSerde}("${fieldName}", "${columnName}", ${superstruct})`
+      ? code`new ${SuperstructSerde}({ ${generateBaseFieldSerdeOpts(dbMetadata, field)} superstruct: ${superstruct} })`
       : columnType === "numeric"
-      ? code`new ${DecimalToNumberSerde}("${fieldName}", "${columnName}")`
-      : code`new ${PrimitiveSerde}("${fieldName}", "${columnName}", "${columnType}")`;
+      ? code`new ${DecimalToNumberSerde}({ ${generateBaseFieldSerdeOpts(dbMetadata, field)} })`
+      : code`new ${PrimitiveSerde}({ ${generateBaseFieldSerdeOpts(dbMetadata, field)} })`;
```

Comment:
Hey @brudil ! Sorry for the delay, finally got to look at this and think a bit...

These changes stemming from my initial "opt-ize" idea all look good--but after seeing them play out with the existing serdes (particularly the KeySerde, which wants different tag names based on usage), I'm wondering if we could ever really guess all of any given serdes' necessary params/opts (b/c it's awkward enough for the existing ones, let alone future ones).

Which, if we can't do that, maybe we need to try and not "guess all necessary opts", but instead just provide a hook to let the user, at codegen time, include in the codegen output whatever they want.

Like besides the _runtime_ serde impl, let the user provide _codegen_ time customization. Something like:

```json
{
  "entities": {
    "Author": {
      "fields": {
         "ipAddress": { "serdeGen": "generateIpSerde@src/codegen.ts" }
...
```

Where `generateIpSerde` would be called at codegen time, and passed the `dbMetadata` and `field` (which in theory would have all metadata the user could ever want, to include in their generated constructor call), i.e.:

```
function generateIpSerde(dbMetadata: ..., field: ...) {
  return imp(`new ${IpSerde}(...whatever params necessary...);
}
```

This would let users completely control their serde's constructor parameters.

_But_ we probably don't need to start there.

Like maybe instead of jumping to a `serdeGen` config flag, we start with just `serde: IpSerde@src/serde.ts` with the API contract being that your serde constructor accepts just fieldName, columnName, dbType (as regular parameters, not opts-style):

```ts
export class IpAddressSerde implements Serde {
  constructor(fieldName: ..., columnName: ..., dbType: ...) {
}
```

Which is:

a) really simple for the common case of "this serde doesn't need any custom opts",

b) we could back away from my initial opts-ize the serde approach since, after flushing it out a bit, it seems obvious in retrospect that fitting all serde constructors into a single box/set-of-opts is unrealistic, _but_

c) with the `serdeGen` idea in our back pocket, we've got some future proofing for, should users ever come back with "the field name + column name + db type API for my serde is not good enough", we can then implement `serdeGen` to give them exact control over the code that calls their serde constructor.

So, apologies for the approach concern, but I think this would be better/simpler for most custom serdes, and it'd leave the internal serdes generally alone/as-is.

Wdyt? Do you mind trying to implement ^?

(And fwiw given I'm more up-to-speed on the problem/approach now, I shouldn't take as long on the next PR review :pray: --thank you!).